### PR TITLE
Add constructors for FixedSize array types (#3879)

### LINKED ISSUE #3879: Array Destructuring APIs
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

As part of https://github.com/apache/arrow-rs/issues/1799 strongly typed ArrayData abstractions were added that allow constructing arrays from their constituent parts, and destructuring arrays back into their parts. Whilst it looks like we won't be moving ahead with a strongly typed ArrayData abstraction, these APIs are powerful and should be ported across to the corresponding Array implementation.

**Describe the solution you'd like**

Each of the various `Array` should have:

* An unsafe `new_unchecked` that constructs it from raw `ScalarBuffer`, `OffsetBuffer`, etc...
* A safe `try_new` method that performs validation and returns a `Result`
* An `into_parts` method that returns the constituent parts that can be fed to `new_unchecked`
* Methods should return `&ScalarBuffer<T>`, `&BooleanBuffer`, etc... instead of `&[T]`, `&[u8]` etc...

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #3879

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/array/fixed_size_list_array.rs [hunk 1]
```diff
@@ -68,6 +69,114 @@ pub struct FixedSizeListArray {
...
+    ///
+    /// # Errors
+    ///
+    /// * `size < 0`
+    /// * `values.len() / size != nulls.len()`
+    /// * `values.data_type() != field.data_type()`
+    /// * `!field.is_nullable() && !nulls.expand(size).contains(values.nulls())`
+    pub fn try_new(
+        field: FieldRef,
+        size: i32,
```
Comment (alamb) [inline arrow-array/src/array/fixed_size_list_array.rs:96]: I think naming this parameter `value_length` to align with the field name would make it easier to understand

File: arrow-data/src/data/mod.rs [hunk 2]
```diff
@@ -1152,40 +1151,17 @@ impl ArrayData {
...
-                                for j in 0..element_len {
-                                    bit_util::set_bit(
-                                        buffer.as_mut(),
-                                        i * element_len + j,
-                                    )
-                                }
-                            }
-                            let mask = buffer.into();
-                            self.validate_non_nullable(Some(&mask), 0, child)?;
+                            let expanded = nulls.expand(element_len);
```
Comment (alamb) [inline arrow-data/src/data/mod.rs:1154]: this is a nice cleanup

**Review by alamb — APPROVED** [review]
👨‍🍳 👌  -- looks like a really nice improvement to me

File: arrow-array/src/array/fixed_size_list_array.rs (hunk 1 above)
Comment (tustvold) [reply arrow-array/src/array/fixed_size_list_array.rs:96]: Given it is a FixedSizeArray and not a FixedValueLengthArray I'm personally inclined to stick with size, but will sleep on it

**Pull request merged**